File: blue-core/src/main/java/blue/time/MeterMap.java
/*
 * blue - object composition environment for csound
 * Copyright (c) 2023 Steven Yi (stevenyi@gmail.com)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published
 * by  the Free Software Foundation; either version 2 of the License or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.LIB.  If not, write to
 * the Free Software Foundation Inc., 59 Temple Place - Suite 330,
 * Boston, MA  02111-1307 USA
 */
package blue.time;

import electric.xml.Element;
import electric.xml.Elements;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.stream.Stream;

/**
 * Manages time signature (meter) changes across a musical timeline.
 * 
 * MeterMap maintains an ordered list of MeasureMeterPair entries, where each
 * entry defines the meter starting at a specific measure number. The meter
 * remains in effect until the next meter change.
 * 
 * Automatically maintains a cache of measure start beats for efficient
 * conversion between measure/beat positions and absolute Csound beats.
 * 
 * Entries are always kept sorted by measure number.
 *
 * @author stevenyi
 */
public class MeterMap {
    
    /**
     * Listener interface for MeterMap changes.
     */
    public interface MeterMapListener {
        void meterMapChanged();
    }

    private final List<MeasureMeterPair> entries = new ArrayList<>();
    private final List<MeterMapListener> listeners = new CopyOnWriteArrayList<>();
    protected double[] measureStartBeats = new double[0];

    public MeterMap() {
        entries.add(new MeasureMeterPair(1, new Meter(4, 4)));
        updateMeasureStartBeats();
    }

    public MeterMap(MeterMap meterMap) {
        for (MeasureMeterPair pair : meterMap.entries) {
            entries.add(new MeasureMeterPair(pair));
        }
        updateMeasureStartBeats();
    }
    
    // ========== Listener Management ==========
    
    public void addListener(MeterMapListener listener) {
        if (listener != null && !listeners.contains(listener)) {
            listeners.add(listener);
        }
    }
    
    public void removeListener(MeterMapListener listener) {
        listeners.remove(listener);
    }
    
    private void fireChanged() {
        updateMeasureStartBeats();
        for (MeterMapListener listener : listeners) {
            listener.meterMapChanged();
        }
    }
    
    // ========== List Operations ==========
    
    public int size() {
        return entries.size();
    }
    
    public boolean isEmpty() {
        return entries.isEmpty();
    }
    
    public MeasureMeterPair get(int index) {
        return entries.get(index);
    }
    
    public void add(MeasureMeterPair pair) {
        entries.add(pair);
        sortEntries();
        fireChanged();
    }
    
    public void set(int index, MeasureMeterPair pair) {
        entries.set(index, pair);
        sortEntries();
        fireChanged();
    }
    
    public void remove(int index) {
        entries.remove(index);
        fireChanged();
    }
    
    public void clear() {
        entries.clear();
        fireChanged();
    }
    
    /**
     * Replaces all entries in this MeterMap with copies from the source.
     * Preserves listeners registered on this object.
     * 
     * @param source the MeterMap to copy entries from
     */
    public void replaceAll(MeterMap source) {
        entries.clear();
        for (MeasureMeterPair pair : source.entries) {
            entries.add(new MeasureMeterPair(pair));
        }
        sortEntries();
        fireChanged();
    }
    
    public Stream<MeasureMeterPair> stream() {
        return entries.stream();
    }
    
    public List<MeasureMeterPair> getEntries() {
        return Collections.unmodifiableList(entries);
    }
    
    private void sortEntries() {
        entries.sort(Comparator.comparingLong(MeasureMeterPair::getMeasureNumber));
    }

    private void updateMeasureStartBeats() {
        measureStartBeats = new double[this.size()];
        
        if (this.isEmpty()) {
            return; // Nothing to update
        }
        
        measureStartBeats[0] = 0;

        for (int i = 1; i < this.size(); i++) {
            var lastMeterEntry = this.get(i - 1);
            var curMeterEntry = this.get(i);

            var measureDur = lastMeterEntry.getMeter().getMeasureBeatDuration();

            var durationOfMeasureRange = measureDur
                    * (curMeterEntry.getMeasureNumber()
                    - lastMeterEntry.getMeasureNumber());

            measureStartBeats[i] = measureStartBeats[i - 1] + durationOfMeasureRange;
        }
    }

    /**
     * Converts a bar/beat position to absolute Csound beats.
     * This is the core conversion method for BBT/BBST/BBF time units.
     * 
     * @param bar the bar (measure) number (1-based)
     * @param beat the beat number within the bar (1-based)
     * @return the equivalent position in Csound beats (without tick fraction)
     * @throws IllegalStateException if MeterMap is empty
     * @throws IllegalArgumentException if bar is before first meter entry
     *         or if beat number exceeds the meter's beat count
     */
    public double barBeatToBeats(long bar, int beat) {
        if (this.isEmpty()) {
            throw new IllegalStateException("MeterMap is empty");
        }

        if (bar < this.get(0).getMeasureNumber()) {
            throw new IllegalArgumentException(
                "Bar " + bar + " is before first meter entry at measure " + 
                this.get(0).getMeasureNumber());
        }

        // Find the meter entry that applies to this bar
        int index = this.size() - 1;
        while (index != 0 && this.get(index).getMeasureNumber() > bar) {
            index--;
        }

        var meterEntry = this.get(index);
        var meter = meterEntry.getMeter();
        
        // Validate beat number doesn't exceed meter's beat count
        if (beat > meter.numBeats) {
            throw new IllegalArgumentException(
                "Beat number " + beat + " exceeds meter " + meter + 
                " which has " + meter.numBeats + " beats per measure");
        }
        
        var measureNumInRange = bar - meterEntry.getMeasureNumber();
        double retVal = measureStartBeats[index];

        retVal += (measureNumInRange * meter.getMeasureBeatDuration()) 
                + ((beat - 1) * (4.0 / meter.beatLength));
        return retVal;
    }

    /**
     * Passthrough conversion for BeatTime (already in Csound beats).
     * 
     * @param beatTime the beat time value
     * @return the Csound beats value
     */
    double toBeats(TimePosition.BeatTime beatTime) {
        return beatTime.getCsoundBeats();
    }

    /**
     * Converts absolute Csound beats to BBT (Bars.Beats.Ticks) time.
     * 
     * @param csoundBeats the absolute beat position to convert
     * @param ppq the PPQ (pulses per quarter note) for tick calculation
     * @return a new BBTTime representing the bar/beat/tick position
     * @throws IllegalStateException if MeterMap is empty
     * @throws IllegalArgumentException if beats is negative
     */
    public TimePosition.BBTTime beatsToBBT(double csoundBeats, int ppq) {
        if (this.isEmpty()) {
            throw new IllegalStateException("MeterMap is empty");
        }
        
        if (csoundBeats < 0) {
            throw new IllegalArgumentException("Beats cannot be negative: " + csoundBeats);
        }
        
        // Find which meter range this beat falls into
        int index = 0;
        for (int i = 1; i < this.size(); i++) {
            if (csoundBeats >= measureStartBeats[i]) {
                index = i;
            } else {
                break;
            }
        }
        
        var meterEntry = this.get(index);
        var meter = meterEntry.getMeter();
        long baseMeasure = meterEntry.getMeasureNumber();
        double beatsFromRangeStart = csoundBeats - measureStartBeats[index];
        
        // Calculate how many complete measures and remaining beats
        double measureDuration = meter.getMeasureBeatDuration();
        long additionalMeasures = (long) (beatsFromRangeStart / measureDuration);
        double remainingBeats = beatsFromRangeStart - (additionalMeasures * measureDuration);
        
        // Convert remaining beats to beat number within measure
        // Beat number starts at 1, and is scaled by meter's beat length
        double beatWithFraction = 1.0 + (remainingBeats / (4.0 / meter.beatLength));
        
        int wholeBeat = (int) beatWithFraction;
        double fractionalBeat = beatWithFraction - wholeBeat;
        int ticks = (int) Math.round(fractionalBeat * ppq);
        
        // Handle tick overflow (ticks == ppq means next beat)
        if (ticks >= ppq) {
            ticks = 0;
            wholeBeat++;
            // Handle beat overflow to next measure
            if (wholeBeat > meter.numBeats) {
                wholeBeat = 1;
                additionalMeasures++;
            }
        }
        
        long bar = baseMeasure + additionalMeasures;
        
        return TimePosition.bbt(bar, wholeBeat, ticks);
    }
    
    /**
     * Converts absolute Csound beats to BBST (Bars.Beats.Sixteenths.Ticks) time.
     * 
     * @param csoundBeats the absolute beat position to convert
     * @param ppq the PPQ (pulses per quarter note) for tick calculation
     * @return a new BBSTTime representing the bar/beat/sixteenth/tick position
     * @throws IllegalStateException if MeterMap is empty
     * @throws IllegalArgumentException if beats is negative
     */
    public TimePosition.BBSTTime beatsToBBST(double csoundBeats, int ppq) {
        TimePosition.BBTTime bbt = beatsToBBT(csoundBeats, ppq);
        return bbt.toBBST(ppq);
    }
    
    /**
     * Converts absolute Csound beats to BBF (Bars.Beats.Fraction) time.
     * 
     * @param csoundBeats the absolute beat position to convert
     * @return a new BBFTime representing the bar/beat/fraction position
     * @throws IllegalStateException if MeterMap is empty
     * @throws IllegalArgumentException if beats is negative
     */
    public TimePosition.BBFTime beatsToBBF(double csoundBeats) {
        if (this.isEmpty()) {
            throw new IllegalStateException("MeterMap is empty");
        }
        
        if (csoundBeats < 0) {
            throw new IllegalArgumentException("Beats cannot be negative: " + csoundBeats);
        }
        
        // Find which meter range this beat falls into
        int index = 0;
        for (int i = 1; i < this.size(); i++) {
            if (csoundBeats >= measureStartBeats[i]) {
                index = i;
            } else {
                break;
            }
        }
        
        var meterEntry = this.get(index);
        var meter = meterEntry.getMeter();
        long baseMeasure = meterEntry.getMeasureNumber();
        double beatsFromRangeStart = csoundBeats - measureStartBeats[index];
        
        // Calculate how many complete measures and remaining beats
        double measureDuration = meter.getMeasureBeatDuration();
        long additionalMeasures = (long) (beatsFromRangeStart / measureDuration);
        double remainingBeats = beatsFromRangeStart - (additionalMeasures * measureDuration);
        
        // Convert remaining beats to beat number within measure
        double beatWithFraction = 1.0 + (remainingBeats / (4.0 / meter.beatLength));
        
        int wholeBeat = (int) beatWithFraction;
        double fractionalBeat = beatWithFraction - wholeBeat;
        int fraction = (int) Math.round(fractionalBeat * 100);
        
        // Handle fraction overflow (fraction == 100 means next beat)
        if (fraction >= 100) {
            fraction = 0;
            wholeBeat++;
            // Handle beat overflow to next measure
            if (wholeBeat > meter.numBeats) {
                wholeBeat = 1;
                additionalMeasures++;
            }
        }
        
        long bar = baseMeasure + additionalMeasures;
        
        return TimePosition.bbf(bar, wholeBeat, fraction);
    }
    
    /**
     * Save MeterMap to XML.
     */
    public Element saveAsXML() {
        Element retVal = new Element("meterMap");
        
        for (MeasureMeterPair pair : entries) {
            retVal.addElement(pair.saveAsXML());
        }
        
        return retVal;
    }
    
    /**
     * Load MeterMap from XML.
     */
    public static MeterMap loadFromXML(Element data) {
        MeterMap meterMap = new MeterMap();
        
        // Clear default entry
        meterMap.clear();
        
        // Load all meter entries
        Elements pairs = data.getElements("measureMeterPair");
        while (pairs.hasMoreElements()) {
            Element pairElement = pairs.next();
            meterMap.add(MeasureMeterPair.loadFromXML(pairElement));
        }
        
        // If empty after loading, add default
        if (meterMap.isEmpty()) {
            meterMap.add(new MeasureMeterPair(1, new Meter(4, 4)));
        }
        
        return meterMap;
    }

}

File: packages/blue-data/src/time/meter-map.ts
/**
 * MeterMap — manages time signature changes across the timeline.
 * Mirrors the Java MeterMap class.
 *
 * Maps measure numbers to Meter objects, providing conversion between
 * bar/beat positions and absolute beat positions.
 */
import { Meter } from './meter';
import { MeasureMeterPair } from './measure-meter-pair';
import { Element } from '../serialization/xml-reader';

/** Default PPQ (pulses per quarter note), matching Java. */
export const DEFAULT_PPQ = 960;

export class MeterMap {
  private entries: MeasureMeterPair[] = [new MeasureMeterPair(1, new Meter(4, 4))];
  private measureStartBeats: number[] = [0.0];
  private listeners: Array<() => void> = [];

  size(): number {
    return this.entries.length;
  }

  get(index: number): MeasureMeterPair {
    return this.entries[index];
  }

  getEntries(): ReadonlyArray<MeasureMeterPair> {
    return this.entries;
  }

  /** Get the meter in effect at the given measure (1-based). */
  getMeterForMeasure(measure: number): Meter {
    for (let i = this.entries.length - 1; i >= 0; i--) {
      if (measure >= this.entries[i].measure) {
        return this.entries[i].meter;
      }
    }
    return this.entries[0].meter;
  }

  /** Get the meter in effect at the given beat position. */
  getMeterAtBeat(beat: number): Meter {
    for (let i = this.entries.length - 1; i >= 0; i--) {
      if (beat >= this.measureStartBeats[i]) {
        return this.entries[i].meter;
      }
    }
    return this.entries[0].meter;
  }

  /** Add a new meter entry at the given measure. */
  add(entry: MeasureMeterPair): void {
    // Remove existing entry at same measure
    this.entries = this.entries.filter((e) => e.measure !== entry.measure);
    this.entries.push(entry);
    this.entries.sort((a, b) => a.measure - b.measure);
    this.updateMeasureStartBeats();
    this.fireListeners();
  }

  /** Replace the entry at the given index. */
  set(index: number, entry: MeasureMeterPair): void {
    this.entries[index] = entry;
    this.entries.sort((a, b) => a.measure - b.measure);
    this.updateMeasureStartBeats();
    this.fireListeners();
  }

  /** Remove all entries. */
  clear(): void {
    this.entries = [];
    this.measureStartBeats = [];
  }

  /** Replace all entries from a source MeterMap. */
  replaceAll(source: MeterMap): void {
    this.entries = source.entries.map(
      (e) => new MeasureMeterPair(e.measure, new Meter(e.meter.numBeats, e.meter.beatLength)),
    );
    this.updateMeasureStartBeats();
    this.fireListeners();
  }

  addListener(listener: () => void): void {
    this.listeners.push(listener);
  }

  private fireListeners(): void {
    for (const listener of this.listeners) {
      listener();
    }
  }

  /**
   * Recompute cached measure start beats.
   * After calling this, measureStartBeats[i] = the absolute beat position
   * where entries[i].measure begins.
   */
  updateMeasureStartBeats(): void {
    this.measureStartBeats = new Array(this.entries.length);
    this.measureStartBeats[0] = 0.0;

    for (let i = 1; i < this.entries.length; i++) {
      const prevEntry = this.entries[i - 1];
      const curEntry = this.entries[i];
      const measuresBetween = curEntry.measure - prevEntry.measure;
      const beatsPerMeasure = prevEntry.meter.getBeatsPerMeasure();
      this.measureStartBeats[i] = this.measureStartBeats[i - 1] + measuresBetween * beatsPerMeasure;
    }
  }

  /**
   * Convert a bar/beat position to absolute Csound beats.
   * Bar and beat are 1-based (matching musical convention).
   *
   * @throws Error if bar < 1, beat < 1, or beat exceeds meter's numBeats
   */
  barBeatToBeats(bar: number, beat: number): number {
    if (this.entries.length === 0) {
      throw new Error('MeterMap is empty');
    }
    if (bar < 1) {
      throw new Error(`Invalid bar number: ${bar} (must be >= 1)`);
    }
    if (beat < 1) {
      throw new Error(`Invalid beat number: ${beat} (must be >= 1)`);
    }

    // Find the meter entry for this bar
    let entryIndex = 0;
    for (let i = this.entries.length - 1; i >= 0; i--) {
      if (bar >= this.entries[i].measure) {
        entryIndex = i;
        break;
      }
    }

    const entry = this.entries[entryIndex];
    const meter = entry.meter;

    // Validate beat doesn't exceed meter's numBeats
    if (beat > meter.numBeats) {
      throw new Error(`Beat ${beat} exceeds meter's numBeats ${meter.numBeats} in bar ${bar}`);
    }

    const measuresFromEntry = bar - entry.measure;
    const startBeat = this.measureStartBeats[entryIndex];
    const beatScale = meter.getBeatScale();

    return startBeat + measuresFromEntry * meter.getBeatsPerMeasure() + (beat - 1) * beatScale;
  }

  /**
   * Convert absolute beats to a BBT position {bar, beat, ticks}.
   * Ticks are in the range [0, ppq).
   */
  beatsToBBT(
    beats: number,
    ppq: number = DEFAULT_PPQ,
  ): { bar: number; beat: number; ticks: number } {
    if (this.entries.length === 0) {
      throw new Error('MeterMap is empty');
    }
    if (beats < 0) {
      throw new Error(`Invalid beats: ${beats} (must be >= 0)`);
    }

    // Find the entry whose measureStartBeats <= beats
    let entryIndex = 0;
    for (let i = this.entries.length - 1; i >= 0; i--) {
      if (beats >= this.measureStartBeats[i]) {
        entryIndex = i;
        break;
      }
    }

    const entry = this.entries[entryIndex];
    const meter = entry.meter;
    const startBeat = this.measureStartBeats[entryIndex];
    const beatsPerMeasure = meter.getBeatsPerMeasure();
    const beatScale = meter.getBeatScale();

    const beatsFromEntry = beats - startBeat;
    const measuresFromEntry = Math.floor(beatsFromEntry / beatsPerMeasure);
    const remainingBeats = beatsFromEntry - measuresFromEntry * beatsPerMeasure;

    // beat is 1-based, compute fractional part as ticks
    const fullBeats = Math.floor(remainingBeats / beatScale);
    const fractionalBeat = remainingBeats - fullBeats * beatScale;
    let ticks = Math.round((fractionalBeat * ppq) / beatScale);

    let bar = entry.measure + measuresFromEntry;
    let beat = fullBeats + 1;

    if (ticks >= ppq) {
      ticks = 0;
      beat += 1;
      if (beat > meter.numBeats) {
        beat = 1;
        bar += 1;
      }
    }

    return { bar, beat, ticks };
  }

  /**
   * Convert absolute beats to a BBST position.
   * Returns {bar, beat, sixteenth, ticks} where beat is 1-based, sixteenth is 1-4.
   */
  beatsToBBST(
    beats: number,
    ppq: number = DEFAULT_PPQ,
  ): { bar: number; beat: number; sixteenth: number; ticks: number } {
    const bbt = this.beatsToBBT(beats, ppq);
    const sixteenthTicks = ppq / 4;
    const sixteenth = Math.floor(bbt.ticks / sixteenthTicks) + 1;
    const ticks = bbt.ticks % sixteenthTicks;
    return { bar: bbt.bar, beat: bbt.beat, sixteenth: Math.min(sixteenth, 4), ticks };
  }

  /**
   * Convert absolute beats to a BBF position.
   * Returns {bar, beat, fraction} where fraction is canonical hundredths.
   * Rounding overflow carries to the next beat, matching Java Blue 2.10.2.
   */
  beatsToBBF(beats: number): { bar: number; beat: number; fraction: number } {
    if (this.entries.length === 0) {
      throw new Error('MeterMap is empty');
    }
    if (beats < 0) {
      throw new Error(`Invalid beats: ${beats} (must be >= 0)`);
    }

    let entryIndex = 0;
    for (let i = this.entries.length - 1; i >= 0; i--) {
      if (beats >= this.measureStartBeats[i]) {
        entryIndex = i;
        break;
      }
    }

    const entry = this.entries[entryIndex];
    const meter = entry.meter;
    const startBeat = this.measureStartBeats[entryIndex];
    const beatsPerMeasure = meter.getBeatsPerMeasure();
    const beatScale = meter.getBeatScale();

    const beatsFromEntry = beats - startBeat;
    const measuresFromEntry = Math.floor(beatsFromEntry / beatsPerMeasure);
    const remainingBeats = beatsFromEntry - measuresFromEntry * beatsPerMeasure;

    const beatWithFraction = 1 + remainingBeats / beatScale;
    let beat = Math.floor(beatWithFraction);
    let fraction = Math.round((beatWithFraction - beat) * 100);
    let bar = entry.measure + measuresFromEntry;

    if (fraction >= 100) {
      fraction = 0;
      beat += 1;
      if (beat > meter.numBeats) {
        beat = 1;
        bar += 1;
      }
    }

    return { bar, beat, fraction };
  }

  equals(other: MeterMap): boolean {
    if (this.entries.length !== other.entries.length) return false;
    for (let i = 0; i < this.entries.length; i++) {
      if (!this.entries[i].equals(other.entries[i])) return false;
    }
    return true;
  }

  hashCode(): number {
    let h = 0;
    for (const entry of this.entries) {
      h = ((h * 31 + entry.measure) >>> 0) | 0;
    }
    return h;
  }

  // ─── XML Serialization ───

  saveAsXML(): Element {
    const elem = new Element('meterMap');
    for (const entry of this.entries) {
      elem.addElement(entry.saveAsXML());
    }
    return elem;
  }

  static loadFromXML(data: Element): MeterMap {
    const map = new MeterMap();
    map.entries = [];

    const entries = data.getElements('measureMeterPair');
    while (entries.hasMoreElements()) {
      const entry = MeasureMeterPair.loadFromXML(entries.next());
      map.entries.push(entry);
    }

    // Default to 4/4 if no entries found
    if (map.entries.length === 0) {
      map.entries.push(new MeasureMeterPair(1, new Meter(4, 4)));
    }

    map.updateMeasureStartBeats();
    return map;
  }
}
